osu.Server.Spectator/Hubs/ScoreProcessedSubscriber.cs

```diff
@@ -0,0 +1,176 @@
+        IConnectionMultiplexer redis)
+    {
+        this.databaseFactory = databaseFactory;
+
+        timer = new Timer(1000);
+        timer.AutoReset = true;
+        timer.Elapsed += (_, _) => Task.Run(purgeTimedOutSubscriptions);
+        timer.Start();
+
+        subscriber = redis.GetSubscriber();
+        subscriber.Subscribe("osu-channel:score:processed", (_, message) => onMessageReceived(message));
+
+        logger = Logger.GetLogger(nameof(ScoreProcessedSubscriber));
+    }
+
+    private void onMessageReceived(string message)
+    {
+        try
+        {
+            var scoreProcessed = JsonConvert.DeserializeObject<ScoreProcessed>(message);
+
+            if (scoreProcessed == null)
+                return;
+
+            if (subscriptions.TryRemove(scoreProcessed.ScoreId, out var subscription))
+            {
+                using (subscription)
+                    subscription.InvokeAsync().Wait();
+            }
+
+            DogStatsd.Increment($"{statsd_prefix}.messages.delivered");
+        }
+        catch (Exception ex)
+        {
+            logger.Add($"Failed to process message {message}", LogLevel.Error, ex);
+            DogStatsd.Increment($"{statsd_prefix}.messages.dropped");
+        }
+    }
+
+    public async Task RegisterForNotificationAsync(string receiverConnectionId, int userId, long scoreToken, ScoreProcessedAsyncCallback callback)
+    {
+        try
+        {
+            using var db = databaseFactory.GetInstance();
+
+            long? scoreId = await db.GetScoreIdFromToken(scoreToken);
+
+            if (scoreId == null)
+            {
+                DogStatsd.Increment($"{statsd_prefix}.subscriptions.dropped");
+                return;
+            }
+
+            var subscription = new ScoreProcessedSubscription(receiverConnectionId, userId, scoreId.Value, callback);
+
+            // because the score submission flow happens concurrently with the spectator play finished flow,
+            // it is theoretically possible for the score processing to complete before the spectator hub had a chance to register for notifications.
+            // to cover off this possibility, check the database directly once.
+            if (await db.IsScoreProcessedAsync(scoreId.Value))
+            {
+                using (subscription)
+                    await subscription.InvokeAsync();
+                DogStatsd.Increment($"{statsd_prefix}.messages.delivered-immediately");
+                return;
+            }
+
+            subscriptions.TryAdd(scoreId.Value, subscription);
+            DogStatsd.Gauge($"{statsd_prefix}.subscriptions.total", subscriptions.Count);
+        }
+        catch (Exception ex)
+        {
+            logger.Add($"Failed to register connection {receiverConnectionId} for info about score {userId}:{scoreToken}", LogLevel.Error, ex);
+            DogStatsd.Increment($"{statsd_prefix}.subscriptions.failed");
+        }
+    }
+
+    private void purgeTimedOutSubscriptions()
```

Comment:
I've added this to avoid a situation where the subscriber gets backlogged due to the score processor not being able to catch up to incoming scores, causing a cascading failure. I'm not sure how realistic of a scenario this is in practice.

Reply:
Going to have to see how this plays out post deploy really.